File: player/src/main/java/com/xyoye/player/commom/receiver/BatteryBroadcastReceiver.java
package com.xyoye.player.commom.receiver;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.os.BatteryManager;

/**
 * 电量改变广播
 *
 * Created by xyoye on 2019/5/7.
 */

public class BatteryBroadcastReceiver extends BroadcastReceiver {
    // 电量改变中
    public static final int BATTERY_STATUS_SPA = 101;
    // 低电量
    public static final int BATTERY_STATUS_LOW = 102;
    // 正常电量
    public static final int BATTERY_STATUS_NOR = 103;
    // 低电量临界值
    private static final int BATTERY_LOW_LEVEL = 15;

    private PlayerReceiverListener listener;

    public BatteryBroadcastReceiver(PlayerReceiverListener listener){
        this.listener = listener;
    }

    @Override
    public void onReceive(Context context, Intent intent) {
        if (intent == null || intent.getAction() == null)
            return;
        // 接收电量变化信息
        if (intent.getAction().equals(Intent.ACTION_BATTERY_CHANGED)) {
            int level = intent.getIntExtra("level", 0);
            int scale = intent.getIntExtra("scale", 100);
            int status = intent.getIntExtra("status", BatteryManager.BATTERY_HEALTH_UNKNOWN);

            // 电量百分比
            int curPower = level * 100 / scale;

            if (status == BatteryManager.BATTERY_STATUS_CHARGING) {
                listener.onBatteryChanged(BATTERY_STATUS_SPA, curPower);
            } else if (curPower < BATTERY_LOW_LEVEL) {
                listener.onBatteryChanged(BATTERY_STATUS_LOW, curPower);
            } else {
                listener.onBatteryChanged(BATTERY_STATUS_NOR, curPower);
            }
        }
    }
}

File: common_component/src/main/java/com/xyoye/common_component/receiver/BatteryBroadcastReceiver.kt
package com.xyoye.common_component.receiver

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.os.BatteryManager

/**
 * Created by xyoye on 2021/11/16.
 */

class BatteryBroadcastReceiver(
    private val battery: (Int) -> Unit,
    private val charging: (Boolean) -> Unit
) : BroadcastReceiver() {
    override fun onReceive(context: Context?, intent: Intent?) {
        when (intent?.action) {
            Intent.ACTION_BATTERY_CHANGED -> {
                val level = intent.getIntExtra(BatteryManager.EXTRA_LEVEL, 0)
                val scale = intent.getIntExtra(BatteryManager.EXTRA_SCALE, 100)
                val battery = level * 100 / scale
                this.battery.invoke(battery)

                val status = intent.getIntExtra(
                    BatteryManager.EXTRA_STATUS,
                    BatteryManager.BATTERY_STATUS_UNKNOWN
                )
                if (status == BatteryManager.BATTERY_STATUS_CHARGING) {
                    charging.invoke(true)
                }
            }
            Intent.ACTION_POWER_CONNECTED -> {
                charging.invoke(true)
            }
            Intent.ACTION_POWER_DISCONNECTED -> {
                charging.invoke(false)
            }
        }
    }
}